File: crc.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class crc{
    public static void divide(List<Integer>n,List<Integer>k,List<Integer>ans,List<Integer>rem){
        int i=0;

        for(int y=0;y<k.size();y++){
            rem.add(n.get(y));
            i++;
        }
        while(i<n.size()){
            if(rem.get(0)==1){
                ans.add(1);
                for(int j=0;j<k.size();j++){
                    int xor = k.get(j) ^ rem.get(j);
                    rem.set(j, xor);
                }
                if(rem.get(0)==0){
                    for(int x=0;x<rem.size()-1;x++){
                        rem.set(x, rem.get(x+1));
                    }
                }
            }
            else if(rem.get(0)==0){
                ans.add(0);
                for(int j=0;j<k.size();j++){
                    int xor = rem.get(j) ^ 0;
                    rem.set(j, xor);
                }
                if(rem.get(0)==0){
                    for(int x=0;x<rem.size()-1;x++){
                        rem.set(x, rem.get(x+1));
                    }
                }
            }
            rem.set(rem.size()-1, n.get(i));
            i++;
           
        }
        if(rem.get(0)==1){
            ans.add(1);
            for(int j=0;j<k.size();j++){
                int xor = k.get(j) ^ rem.get(j);
                rem.set(j, xor);
            }
            
        }
        else if(rem.get(0)==0){
            ans.add(0);
            for(int j=0;j<k.size();j++){
                int xor = rem.get(j) ^ 0;
                rem.set(j, xor);
            }
          
        }
        System.out.println("Syndrome: "+ rem.toString());
    }
    public static void main(String[] args) {
        List<Integer>n = new ArrayList<>();
        List<Integer>k= new ArrayList<>(Arrays.asList(1,0,1,0,1));
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter data :");
        String data = sc.nextLine();
        //convert to ascii to binary
        int v=0;
        for(int i=0;i<data.length();i++){
             v = v+ (int)(data.charAt(i));
        }
        String binary = Integer.toBinaryString(v);


        for(int i=0;i<binary.length();i++){
            n.add(binary.charAt(i)-'0');
        }

        for(int i=0;i<k.size()-1;i++){
            n.add(0);
        }

        System.out.println("data: "+n.toString());
        System.out.println("Generator: "+k.toString());
        
        List<Integer>ans = new ArrayList<>();
        List<Integer>rem = new ArrayList<>();
        divide(n, k, ans, rem);
        
        int lengthn = n.size();
        int lengthen = k.size();
        int z=0;
        for(int f=lengthn-1;f>lengthn-lengthen;f--){
            n.set(f, rem.get(z));
            z++;

        }
        System.out.println("Do you want to add error");
        System.out.println("1-->Yes");
        System.out.println("2-->No");
        int ch = sc.nextInt();
        if(ch==1){
            int r = (int) (Math.random()*lengthen);
            // System.out.println("Bit changed position "+(lengthn-r));
            if(n.get(r)==0) n.set(r, 1);
            else if (n.get(r)==1) n.set(r, 0);
            List<Integer>ans1 = new ArrayList<>();
            List<Integer>rem1 = new ArrayList<>();

            divide(n, k, ans1, rem1);
            System.out.println("Error detected");
        }
        if(ch==2){
            List<Integer>ans2 = new ArrayList<>();
            List<Integer>rem2 = new ArrayList<>();
            divide(n, k, ans2, rem2);

        }


        sc.close();



    }
}

File: crc.cpp
#include <iostream>
#include <string>
#include <bitset>
#include <cstdlib>
#include <ctime>
using namespace std;

string crcdiv(string data, const string &gen) {
    int genLen = gen.size();
    string syndrome = data.substr(0, genLen);

    for (int i = genLen; i <= data.size(); i++) {
        if (syndrome[0] == '1') {
            for (int j = 0; j < genLen; j++)
                syndrome[j] = (syndrome[j] == gen[j]) ? '0' : '1';
        } else {
            for (int j = 0; j < genLen; j++)
                syndrome[j] = (syndrome[j] == '0') ? '0' : '1';
        }
        syndrome.erase(0, 1);
        if (i < data.size())
            syndrome.push_back(data[i]);
    }
    return syndrome;
}

string crcencode(string data, const string &gen) {
    data.append(gen.size() - 1, '0');
    return crcdiv(data, gen);
}

string crccheck(string codeword, const string &gen) {
    return crcdiv(codeword, gen);
}

int main() {
    srand(time(0));

    string generator;
    cout << "Enter generator polynomial (binary): ";
    cin >> generator;

    int zeros;
    cout << "Enter no. of zeroes to augment: ";
    cin >> zeros;

    cin.ignore();
    string input;
    cout << "Enter string: ";
    getline(cin, input);

    cout << "\n<Sender's Side>\n";

    // Process each character separately
    for (char c : input) {
        string binChar = bitset<7>(c).to_string();

        cout << "\nCharacter: " << c << endl;
        cout << "ASCII value: " << (int)c << endl;
        cout << "Binary value: " << binChar << endl;

        string remainder = crcencode(binChar, generator);
        cout << "CRC remainder: " << remainder << endl;

        string codeword = binChar + remainder;
        cout << "Codeword to be sent = " << codeword << endl;

        cout << "\n<Receiver's Side>\n";
        char choice;
        cout << "Do you want to induce error (y/n)? ";
        cin >> choice;

        string received = codeword;

        if (choice == 'y' || choice == 'Y') {
            // flip a random bit in this character’s codeword
            int bitPos = rand() % received.size();
            received[bitPos] = (received[bitPos] == '0') ? '1' : '0';
            cout << "\nRandom error generated!" << endl;
            cout << "Received Codeword = " << received << endl;

            string syndrome = crccheck(received, generator);
            cout << "Syndrome after error: " << syndrome << endl;
            cout << "Therefore this is error → CRC failed.\n";
        } else {
            cout << "\nNo error induced." << endl;
            cout << "Received Codeword = " << received << endl;

            string syndrome = crccheck(received, generator);
            cout << "Syndrome = " << syndrome << endl;
            if (stoi(syndrome, nullptr, 2) == 0) {
                cout << "There is no error." << endl;
                string dataword = received.substr(0, received.size() - (generator.size() - 1));
                cout << "Fetch the dataword: " << endl;
                cout << "Binary no: " << dataword << endl;
                char decoded = static_cast<char>(bitset<7>(dataword).to_ulong());
                cout << "Alphabet: " << decoded << endl;
            }
        }
    }

    return 0;
}
